src/man1/q4wine-helper.1
.TH "q4wine-helper" "1" "13 Apr 2025" "Q4Wine 1.4 Manual" "Q4Wine Manual"
.SH "NAME"
q4wine-helper \- run Windows program in Q4Wine helper environment.
.SH "SYNOPSIS"
\fB q4wine-helper\fR [\fBKEYs\fR] ...
.br
.SH "DESCRIPTION"
\fBq4wine-helper\fR is a console utility for Q4Wine which helps to handle Wine application exit status and its stdout\\stderr output logging.
.br
.SH "OPTIONS"
.RB "A summary of options is included below."
.TP
.BR "\-\-prefix"
sets current prefix name
.TP
.BR "\-\-wine-debug"
sets WINEDEBUG variable
.TP
.BR "\-\-console"
run with output in console
.TP
.BR "\-\-display"
sets DISPLAY variable
.TP
.BR "\-\-nice"
sets program niceness. See \fBman nice(1)\fR for details
.TP
.BR "\-\-desktop"
sets program virtual desktop resolution
.TP
.BR "\-\-program-bin"
sets program binary
.TP
.BR "\-\-program-args"
sets program args
.TP
.BR "\-\-wrkdir"
sets program working directory
.TP
.BR "\-\-program-lang"
sets program LANG variable
.TP
.BR "\-\-override"
sets WINEDLLOVERRIDES variable
.SH "EXIT STATUS"
.B 0
if OK or
.B \-1
if serious troubles
.SH "FILES"
.TP
.I /usr/bin/q4wine
The
.B Q4Wine
main gui program
.TP
.I /usr/bin/q4wine-cli
The
.B Q4Wine
command line interface program
.TP
.I /usr/bin/q4wine-helper
The
.B Q4Wine
runner helper program
.TP
.I /usr/lib/q4wine
Directory containing
.B Q4Wine's
shared libraries
.TP
.I $HOME/.config/q4wine
Directory containing Q4Wine config, icons, database and temporary files.

.SH "AUTHORS"
.SS Thanks to all:
Q4Wine is available thanks to the work of many to all people, who were
involved in development process, testing, bugreporting, packaging or
translation creation. For a full listing of the authors, please see
the file
.B AUTHORS
in the top-level directory of the source distribution.

.SS Created by:
.TP
Oleksii S. Malakhov (brezerk) <\fBbrezerk at gmail.com\fR>
Project maintainer, lead developer, web site designer and developer,
project documentation, support, etc. ;)

.SH "COPYRIGHT"
Q4Wine can be distributed under the terms of the GPLv3 license.
A copy of the license is in the file
.B LICENSE
in the top-level directory of the source distribution.

.SH "AVAILABILITY"
The most recent
.B public
version of wine can be downloaded from
.I https://q4wine.brezblock.org.ua/download/

The latest snapshot of the code may be obtained via
.B GIT
for information on how to do this, please see
.I https://q4wine.brezblock.org.ua/download/#git

Q4Wine, development headquarters, is at
.I https://q4wine.brezblock.org.ua
This website contains a great deal of information about Q4Wine.

.SH "BUGS"
If you find a bug in Q4Wine please report it to
<\fBbrezerk at gmail.com\fR> or directly to Q4Wine
bug tracker
.I https://sourceforge.net/projects/q4wine/
.PP
.SH "SEE ALSO"
.PP
\fBq4wine\fR(1),
\fBq4wine-cli\fR(1),
\fBwine\fR(1),
\fBwineserver\fR(1)\&


ChangeLog
Version 1.3.13
 Fixed:
  - Prefix Import feature is not working as expected [BUG-156];
  - Misspelled Color Theme Settings [BUG-155];
  - CMake warning [BUG-158];
  - XDG Base Directory Specification support [BUG-178];
  - Use LC_ALL instead of LANG to force override applciation locale [BUG-165];

Version 1.3.12
 Updated:
  - Afrikaans translation;
  - Japanese translation;
  - Ukrainian translation;
  - Russian translation;
  - Traditional Chinese translation;
  - German translation;
 Add:
  - Add an option to show / hide "Args" parameter [BUG-114];
  - Generate StartupWMClass values in created .desktop files [BUG-153];
  - Added a feature to keep the terminal open when running winetricks so users can see the output.
 Fixed:
  - Fix parsing of env causing issues with wine kill [BUG-143];
  - Error: wrong or broken XML data. Try again later. [BUG-146];
  - Fixed a bug where /etc/mtab was not being read and resulted in being unable to unmount iso images.
  - Fixed a bug where the temporary director for generating icons wasn't being deleted. Refactored this to use QTemporaryDir.
  - Added workaround for st terminal, similar to konsole.
  - Fixed warnings: Qt deprecated function useage;
  - Fixed warnings: use of old-style cast

Version 1.3.11
 Fixed:
  - Display of Valve Proton in Processes view [BUG-138];
  - Improve all dialogs of choosing files [BUG-139];
  - Allow Stop Wine action to kill Valve Proton in Processes view [BUG-138];
 Updated:
  - Code cleanup: drop Qt4 support;

Version 1.3.10
 Fixed:
  - cppcheck: Uninitialized struct member: item.provider_id;

 Add:
  - cleanup.sh: to get rid of all non-release stuff;

Version 1.3.9
 Fixed:
  - Fixed build with Qt >= 5.11 [BUG-125];
  - Error: wrong or broken XML data. Try again later. [BUG-132]

 Added:
  - Custom virtual desktop resolution size [BUG-131];

Version 1.3.8
 Updated:
  - gui_sudo missing in Ubuntu 18.04 [BUG-120];

 Fixed:
  - winetricks: Proxy settings are used even if proxy type set to 'none';

Version 1.3.7
 Updated:
  - Wine stock icons to be organised through Q4Wine [BUG-117];
  - Add gnome-terminal into terminal detection logic [BUG-118];
 Added:
  - French translation (thx 2 seigneurfuo);

Version 1.3.6
 Updated:
  - Added indication of applications which was installed through winetricks [BUG-106][BUG-111];
  - Default Wine Desktop Integartion path settings for new prefixes [BUG-106];
  - Added option to run custom scripts with fake drive settings after creation of fake drive [BUG-106];
  - Qt5 build enabled by default;
  - Taiwan translation;
  - Russian translation;
  - Ukrainian translation;
  - Spanish translation;
  - Czech translation;
  - Bind Plus and Minus keys for Program tab for +\- zoom respectively [BUG-112];
  - Bind Plus and Minus keys for Config tab for +\- zoom respectively [BUG-112];
  - Allow to change Config and Program Icon sizes independently  [BUG-112];
 Documented:
  - q4wine prevents user from running winecfg from KRunner even after uninstalling [BUG-98];
 Fixed:
  - Static code analysis: errors and warnings [BUG-116];
  - Error: wrong or broken XML data. Try again later. [BUG-114];

Version 1.3.5
 Updated:
  - Italian translation;
  - Ukrainian translation;
  - Russian translation;
 Fixed:
  - Removed extra Doxygen files to reduce artifact size;

Version 1.3.4
 Added:
  - Add WINE env var to console environment [BUG-91];
 Fixed:
  - Icons are not loaded properly with Qt 5.7.x;
  - Console launch environment uses dead setting fields [BUG-90];
  - Q4Wine file selectors filters files with upper case extension [BUG-94][QTBUG-51712];
  - Refactor winetrkics parser: use direct parser + transactions;
  - Links imported from Windows desktop do not work [BUG-103];
  - Import from Windows desktop is not working in some cases [BUG-103];
  - Processes tab does not detect some wine processes correctly [BUG-100];

Version 1.3.3
 Fixed:
  - Problem opening file dialog to mount ISO [BUG-86];

Version 1.3.2
 Fixed:
  - [regression, bisected] Icon information is not displayed correctly [BUG-80];
  - Winetricks does not update [BUG-77];
  - First startup wizard (quick mount profile): profile strings are initially empty if fuseiso is not installed [BUG-81];
  - Download (latest) Winetricks from GitHub [BUG-82];
  - Correct proxy-related environment variables in Winetricks plugin [BUG-83];
  - Use env command in the command line to download Winetricks [BUG-83];
  - Support proxy username/password in Winetricks plugin [BUG-83];
  - Treat aarch64 as 64 bit platform and put libraries in lib64 directory [BUG-85];

Version 1.3.1
 Added:
  - Allow to execute and create icons for *.bat and *.msi files [BUG-69];
  - Added Chinese (China) translation (thx 2 Mingye Wang and Mingcong Bai);
  - Added Japanese translation (thx 2 Masanori Kakura);

 Fixed:
  - Renamed 'New' menu option for icon and prefix view to 'New icon' and 'New
  folder' respectively to avoid confusion [BUG-69];
  - Icon description is not exported into *.desktop file [BUG-69];
  - Winetrick's item description label doesn't wrap properly [BUG-71];
  - Qt5: detailed view item selection duplicate [BUG-75];
  - q4wine-cli does not accept extra app arguments [BUG-78];
  - Translation files moved from i18n to l10n;
  - Hide deprecated prefix options if values are not set;

Version 1.3
 Added:
  - Taiwan translation (thx 2 Jeff Huang);
  - Use icons from the system icon theme where possible [BUG-22];
  - Replace default raster icons with SVG analogs [BUG-21];
  - Extend App icons zoom sizes up to 512px [BUG-21];
  - Allow to choose *.svg and *.svgz images for icons [BUG-21];
  - Add an app option to switch default/ambiance-light/ambiance-dark icon [BUG-21];
  - Use ISO country codes for translation files (required for: pootle, gentoo ebuild);
  - Use native language name for language selection;
 Fixed:
  - WineLibs paths are no longer strictly required [BUG-68];
  - Process explorer is broken [BUG-70];

Version 1.2-r2
 Added:

src/q4wine-helper/q4wine-helper.cpp
/***************************************************************************
 *   Copyright (C) 2008-2025 by Oleksii S. Malakhov <brezerk@gmail.com>    *
 *                                                                         *
 *   This program is free software: you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation, either version 3 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
 *                                                                         *
 ***************************************************************************/

#include "q4wine-helper.h"

int main(int argc, char *argv[]) {
  QCoreApplication app(argc, argv);
  QTextStream QErr(stderr);

  //! This is need for libq4wine-core.so import;
  typedef void *CoreLibPrototype(bool);
  CoreLibPrototype *CoreLibClassPointer;
  std::unique_ptr<corelib> CoreLib;
  QLibrary libq4wine;

  // Loading libq4wine-core.so
#ifdef RELEASE
  libq4wine.setFileName(_CORELIB_PATH_);
#else
  libq4wine.setFileName(QString("%1/q4wine-lib/libq4wine-core")
                            .arg(QString::fromUtf8(APP_BUILD)));
#endif

  if (!libq4wine.load()) {
    libq4wine.load();
  }

  // Getting corelib class pointer
  CoreLibClassPointer =
      reinterpret_cast<CoreLibPrototype *>(libq4wine.resolve("createCoreLib"));
  CoreLib.reset(static_cast<corelib *>(CoreLibClassPointer(false)));

  if (!CoreLib.get()) {
    QErr << "[EE] Cannot load shared library." << Qt::endl;
    return -1;
  }

  DataBase db;
  if (!db.checkDb()) {
    QErr << "[EE] Cannot initialize database engine." << Qt::endl;
    return -1;
  }

  QTranslator qtt;

#ifdef RELEASE
#ifdef _OS_DARWIN_
  QString l10nPath = QString("%1/%2.app/Contents/l10n")
                         .arg(QDir::currentPath())
                         .arg(APP_SHORT_NAME);
#else
  QString l10nPath = QString("%1/share/%2/l10n")
                         .arg(QString::fromUtf8(APP_PREF))
                         .arg(APP_SHORT_NAME);
#endif
#else
  QString l10nPath = QString("%1/l10n").arg(QString::fromUtf8(APP_BUILD));
#endif

  qtt.load(CoreLib->getTranslationLang(), l10nPath);
  app.installTranslator(&qtt);

  if (!CoreLib->isConfigured()) {
    QErr << "[EE] App is not configured! Rerun wizard, or delete Q4Wine broken "
            "config files."
         << Qt::endl;
    return -1;
  }

  QTextStream Qcout(stdout);
  WineObject wineObject;

  if (argc == 1) {
    app.arguments().append("-h");
    argc++;
  }

  for (int i = 1; i < argc; i++) {
    qDebug() << app.arguments().at(i);
    if ((app.arguments().at(1) == "--version") or
        (app.arguments().at(1) == "-v")) {
      Qcout << QString("%1-helper %2").arg(APP_SHORT_NAME).arg(APP_VERS)
            << Qt::endl;
      Qcout << QString(
                   "Copyright (C) 2008-2025 by Oleksii S. Malakhov "
                   "<brezerk@gmail.com>")
            << Qt::endl;
      Qcout << QString(
                   "License GPLv3+: GNU GPL version 3 or later "
                   "<http://gnu.org/licenses/gpl.html>.")
            << Qt::endl;
      Qcout << QObject::tr(
                   "This is free software: you are free to change and "
                   "redistribute it.")
            << Qt::endl;
      Qcout << QObject::tr(
                   "There is NO WARRANTY, to the extent permitted by law.")
            << Qt::endl;
      CoreLib->getBuildFlags();
      Qcout << QObject::tr("Author: %1.").arg("Oleksii S. Malakhov")
            << Qt::endl;
      return 0;
    } else if (app.arguments().at(i) == "--prefix") {
      i++;
      if (i < argc) wineObject.setPrefix(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--wine-debug") {
      i++;
      if (i < argc) wineObject.setProgramDebug(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--console") {
      i++;
      if (i < argc) wineObject.setUseConsole(app.arguments().at(i).toInt());
    } else if (app.arguments().at(i) == "--display") {
      i++;
      if (i < argc) wineObject.setProgramDisplay(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--nice") {
      i++;
      if (i < argc) wineObject.setProgramNice(app.arguments().at(i).toInt());
    } else if (app.arguments().at(i) == "--desktop") {
      i++;
      if (i < argc) wineObject.setProgramDesktop(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--program-bin") {
      i++;
      if (i < argc) wineObject.setProgramBinary(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--program-args") {
      i++;
      if (i < argc) wineObject.setProgramArgs(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--wrkdir") {
      i++;
      if (i < argc) wineObject.setProgramWrkdir(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--program-lang") {
      i++;
      if (i < argc) wineObject.setProgramLang(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--override") {
      i++;
      if (i < argc) wineObject.setProgramOverride(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--prerun") {
      i++;
      if (i < argc) wineObject.setPreRun(app.arguments().at(i));
    } else if (app.arguments().at(i) == "--postrun") {
      i++;
      if (i < argc) wineObject.setPostRun(app.arguments().at(i));
    } else {
      Qcout << QObject::tr("Usage:") << Qt::endl;
      Qcout << QObject::tr("  %1-helper [KEYs]...").arg(APP_SHORT_NAME)
            << Qt::endl;
      Qcout << QObject::tr(
                   "Console utility for Q4Wine which helps to handle Wine "
                   "application exit status and its stdout/stderr output "
                   "logging.")
            << Qt::endl
            << Qt::endl;
      Qcout << QObject::tr("KEYs list:") << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --prefix"
            << QObject::tr("sets the current prefix name") << qSetFieldWidth(0)
            << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --wine-debug"
            << QObject::tr("sets WINEDEBUG variable") << qSetFieldWidth(0)
            << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --console"
            << QObject::tr("run with output in console") << qSetFieldWidth(0)
            << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --display"
            << QObject::tr("sets DISPLAY variable") << qSetFieldWidth(0)
            << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --nice"
            << QObject::tr("sets program niceness") << qSetFieldWidth(0)
            << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --desktop"
            << QObject::tr("sets program virtual desktop resolution")
            << qSetFieldWidth(0) << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --program-bin"
            << QObject::tr("sets program binary") << qSetFieldWidth(0)
            << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --program-args"
            << QObject::tr("sets program args") << qSetFieldWidth(0)
            << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --wrkdir"
            << QObject::tr("sets program working directory")
            << qSetFieldWidth(0) << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --program-lang"
            << QObject::tr("sets program LANG variable") << qSetFieldWidth(0)
            << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --override"
            << QObject::tr("sets WINEDLLOVERRIDES variable")
            << qSetFieldWidth(0) << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --prerun"
            << QObject::tr("execute script before program run")
            << qSetFieldWidth(0) << Qt::endl;
      Qcout << qSetFieldWidth(25) << Qt::left << "  --postrun"
            << QObject::tr("execute script after program run")
            << qSetFieldWidth(0) << Qt::endl;
      Qcout << Qt::endl;
      Qcout << QObject::tr("Report %1 bugs to %2")
                   .arg(APP_SHORT_NAME)
                   .arg(APP_BUG_EMAIL)
            << Qt::endl;
      Qcout << QObject::tr("%1 homepage: <%2>")
                   .arg(APP_SHORT_NAME)
                   .arg(APP_WEBSITE)
            << Qt::endl;
      Qcout << QObject::tr(
                   "General help using GNU software: "
                   "<http://www.gnu.org/gethelp/>")
            << Qt::endl;
      return 0;
    }
  }

  return wineObject.run();
}
